hermes-management/src/main/java/pl/allegro/tech/hermes/management/domain/topic/TopicService.java

```diff
@@ -190,18 +140,32 @@ public void removeTopicWithSchema(Topic topic, RequestUser removedBy) {
         removeTopic(topic, removedBy);
     }
 
-    private void removeSchema(Topic topic) {
-        if (AVRO.equals(topic.getContentType()) && topicProperties.isRemoveSchema()) {
-            schemaService.getSchema(topic.getQualifiedName()).ifPresent(s ->
-                    schemaService.deleteAllSchemaVersions(topic.getQualifiedName()));
+    private void removeRelatedSubscriptions(Topic topic, RequestUser removedBy) {
+        List<Subscription> subscriptions = subscriptionRepository.listSubscriptions(topic.getName());
+
+        if (subscriptions.isEmpty()) {
+            return;
         }
+
+        ensureSubscriptionsHaveAutoRemove(subscriptions, topic.getName());
+        subscriptions.forEach(sub -> removeSubscription(topic.getName(), sub.getName(), removedBy));
     }
 
-    private void removeTopic(Topic topic, RequestUser removedBy) {
-        multiDcExecutor.executeByUser(new RemoveTopicRepositoryCommand(topic.getName()), removedBy);
-        multiDCAwareService.manageTopic(brokerTopicManagement -> brokerTopicManagement.removeTopic(topic));
-        auditor.objectRemoved(removedBy.getUsername(), topic);
-        topicOwnerCache.onRemovedTopic(topic);
+    private void ensureSubscriptionsHaveAutoRemove(List<Subscription> subscriptions, TopicName topicName) {
+        boolean anySubscriptionWithoutAutoRemove = subscriptions.stream()
+                .anyMatch(sub -> !sub.isAutoDeleteWithTopicEnabled());
+
+        if (anySubscriptionWithoutAutoRemove) {
+            logger.info("Cannot remove topic due to connected subscriptions");
+            throw new TopicNotEmptyException(topicName);
+        }
+    }
+
+    private void removeSubscription(TopicName topicName, String subscriptionName, RequestUser removedBy) {
+        auditor.beforeObjectRemoval(removedBy.getUsername(), Subscription.class.getSimpleName(), subscriptionName);
+        multiDcExecutor.executeByUser(new RemoveSubscriptionRepositoryCommand(topicName, subscriptionName), removedBy);
+        auditor.objectRemoved(removedBy.getUsername(), subscriptionName);
```

Comment:
Yes, that's right. I wanted to avoid the dependency cycle. 

In my opinion, an additional class like `SubscriptionRemover` is a good idea 👌 
So additionally, I'll extract the logic related to removal from the `SubscriptionService`